cdk/src/open-data-platform/data-plane/data-import/handler-factory.ts

```diff
@@ -48,59 +82,112 @@ export const geoJsonHandlerFactory = (
       ]);
 
       pipeline
-        .on('data', async (batch: any[]) => {
-          console.log(`Processing batch. ${numberRows} rows have been read so far.`);
-          // Pause reads while inserting into db.
-          pipeline.pause();
-          await Promise.all(
-            batch.map(async (row: any) => {
-              // Stop processing if at the limit.
-              if (numberRows >= rowLimit) {
-                console.log('Stopping after row write limit:', rowLimit);
-                pipeline.destroy();
-                return;
-              }
-              // Multiple rows may be processed concurrently, so increment and check the limit
-              // before starting the processing step.
-              ++numberRows;
-              await callback(row, db);
+        .on('data', async (rows: any[]) => {
+          const id = bactchId++;
+
+          // Skip rows up to offset.
+          if (skippedRowCount + batchSize <= rowOffset) {
+            skippedRowCount += batchSize;
+            console.log(
+              `batch${id}: Skipping batch of ${rows.length} rows. ${skippedRowCount}/${rowOffset} offset rows have been skipped so far.`,
+            );
+            return;
+          }
+
+          // Start processing the row.
+          console.log(`batch${id}: Processing batch of ${rows.length} rows.`);
+          const promise = callback(rows, db);
+          promises.push(promise);
+
+          // Reprocess errored rows individually. This may catch transient errors, or at least let
+          // other rows in the batch succeed.
+          promise.catch((_) =>
+            rows.forEach((row) => {
+              console.log(
+                `batch${id}: Reprocessing errored row individually.`,
+                row.value?.properties,
+              );
+              const rePromise = callback([row], db);
+              rePromise.catch((reason) => console.log(`batch${id}: Reprocessing failed`, reason));
+              promises.push(rePromise);
             }),
-          );
-          pipeline.resume();
+          ); // Supress unhandled rejections here.
+
+          try {
+            // Await this batches so we know whether to increment the row count.
+            // This does not block other batches from being processed in parallel.
```

Comment:
is this just so that the logging prints accurate # of rows processed? IIUC this might be causing the processing of more rows than specified since we await to increment processedRowCount while other batches are initiated.

Reply:
Good point. I'll try to address that. My main concern was that I don't want to count rows before they're inserted in case they ultimately fail. But that's minor.

Reply:
I added a variable `inProcessRowCount`, which stores the number of rows being processed. This means we can do a check before processing a new batch rather than after.